docs/reference/remap2/features/progressive_type_safety.cue

```diff
@@ -0,0 +1,74 @@
+remap2: features: progressive_type_safety: {
+	title: "Progressive type safety"
+	description: """
+		VRL implements _progressive_ type safety, erroring at compilation-time if a type mismatch is detected. This
+		largely contributes to VRL's safety principle, ensuring scripts work as expected post-compilation.
+		"""
+
+	characteristics: {
+		errors: {
+			title: "Errors"
+			description: """
+				If VRL detects a type mismatch, it will produce a user-friendly compile-time error. For example, given
+				this VRL script:
+
+				```vrl
+				foo, err = 5
+				```
+
+				VRL will produce the following compile-time error:
+
+				```
+				error: unneeded error assignment
+				  ┌─ :1:1
+				  │
+				1 │ foo, err = 5;
+				  │ ^^^^^^^^   - because this expression cannot fail
+				  │ │
+				  │ this error assignment is unneeded
+				  │
+				  = hint: assign to "foo", without assigning to "err"
+				  = see language documentation at: https://vector.dev
+				```
+				"""
+		}
+		progressive: {
+			title: "Progressivenes"
+			description: """
+				VRL's type safety is _progressive_, meaning it will implement type safety for any value for which it
+				knows the type. Because obervability data can be quite unpredictable, it's not always known which
+				type a field might be, hence the _progressive_ nature of VRL's type safety. As VRL scripts are
+				evaluated, type information is built up and used at compile-time to enforce type safety. Let's look
+				at an example:
+
+				```vrl
+				.foo # any
+				.foo, err = downcase(.foo) # string or null
+				.foo = upcase(.foo) # string
```

Comment:
This still isn't accurate because `.foo` can be `null` and thus `upcase` can fail (see my previous comment).

Reply:
Would this be resolved if it was wrapped in an `if` statement?

```
.foo # any
.foo, err = downcase(.foo) # string or null
if err != null {
	.foo = upcase(.foo) # string
}
```

Being able to infer types like this will really help relax error handling.

Reply:
I already mentioned that in my previous (resolved) comment, but no, that doesn’t work yet. There is a tracking issue to add support for this however.

Reply:
See: https://github.com/timberio/vector/pull/6089#discussion_r559045197

Reply:
Which issue is that? I'd like to prioritize it before 0.12.